themes/default/content/blog/resource-methods-for-pulumi-packages/index.md

```diff
@@ -0,0 +1,609 @@
+---
+title: "Introducing Resource Methods for Pulumi Packages"
+date: 2021-10-11T08:00:00-07:00
+draft: true
+meta_desc: It's now possible to provide resource methods from Pulumi Packages
+meta_image: meta.png
+authors:
+    - justin-vanpatten
+tags:
+    - packages
+---
+
+It's now possible to provide resource methods from Pulumi Packages. Resource methods are similar to functions, but instead of being exposed as top-level functions in a module, methods are exposed as methods on a resource class. This allows for a more object oriented approach to exposing functionality -- operations performed by a resource (that potentially use the resource's state) can now be exposed as methods on the resource. Resource methods can be implemented once, in your language of choice, and made available to users in all Pulumi languages.
+
+<!--more-->
+
+When authoring component resources, it's often useful to provide additional functionality through methods on the component. For example, the `Cluster` component in the [`eks`]({{<relref "/docs/reference/pkg/eks">}}) package has a [`getKubeconfig`](https://github.com/pulumi/pulumi-eks/blob/700d73e961976e58762cb9c723ad2fa838052f46/nodejs/eks/cluster.ts#L1482) method that can be used to generate a kubeconfig for authentication with the cluster that does not use the default AWS credential provider chain, but is instead scoped based on the passed-in arguments. Until now, that method has only been available from JavaScript/TypeScript (the language the `Cluster` component was written in). With the new support for resource methods for Pulumi Packages, we can make this method available to all the Pulumi languages, which is exactly what we've done in [#628](https://github.com/pulumi/pulumi-eks/pull/628).
```

Comment:
I'm on PTO this week, but if we get https://github.com/pulumi/pulumi-eks/pull/628 reviewed/merged and released (v0.34.0?), we could point folks to the released version.

Reply:
Relrefs need spaces in there to get generated correctly. 

```suggestion
When authoring component resources, it's often useful to provide additional functionality through methods on the component. For example, the `Cluster` component in the [`eks`]({{< relref "/docs/reference/pkg/eks" >}}) package has a [`getKubeconfig`](https://github.com/pulumi/pulumi-eks/blob/700d73e961976e58762cb9c723ad2fa838052f46/nodejs/eks/cluster.ts#L1482) method that can be used to generate a kubeconfig for authentication with the cluster that does not use the default AWS credential provider chain, but is instead scoped based on the passed-in arguments. Until now, that method has only been available from JavaScript/TypeScript (the language the `Cluster` component was written in). With the new support for resource methods for Pulumi Packages, we can make this method available to all the Pulumi languages, which is exactly what we've done in [#628](https://github.com/pulumi/pulumi-eks/pull/628).
```